botorch/acquisition/proximal.py

```diff
@@ -0,0 +1,109 @@
+    """A wrapper around AquisitionFunctions to add proximal weighting of the
+    acquisition function. Acquisition function is weighted via a squared exponential
+    centered at the last training point, with varying lengthscales corresponding to
+    'proximal_weights'. Can only be used with acqusition functions based on single
+    batch models.
+
+    Small values of proximal_weights corresponds to strong biasing towards recently
+    observed points, which smoothes optimization with a small potential decrese in
+    convergence rate.
+
+    Example:
+        >>> model = SingleTaskGP(train_X, train_Y)
+        >>> EI = ExpectedImprovement(model, best_f=0.0)
+        >>> proximal_weights = torch.ones(d)
+        >>> EI_proximal = ProximalAcquisitionFunction(EI, proximal_weights)
+        >>> eip = EI_proximal(test_X)
+    """
+
+    def __init__(
+        self,
+        acq_function: AcquisitionFunction,
+        proximal_weights: Tensor,
+    ) -> None:
+        r"""Derived Acquisition Function weighted by proximity to recently
+        observed point.
+
+        Args:
+            acq_function: The base acquisition function, operating on input tensors
+                of feature dimension `d`.
+
+            proximal_weights: A `d'-dim tensor used to bias locality
+                along each axis.
+
+        """
+        Module.__init__(self)
+
+        self.acq_func = acq_function
+
+        self.register_buffer("proximal_weights", proximal_weights)
+
+        # check model for train_inputs and single batch
+        if not hasattr(self.acq_func.model, "train_inputs"):
+            raise UnsupportedError("Acqusition function model must have train_inputs.")
+
+        if self.acq_func.model.batch_shape != torch.Size([]):
+            raise UnsupportedError(
+                "Proximal acqusition function requires batch size " "of one."
+            )
+
+        # check to make sure that weights match the training data shape
+        if (
+            self.proximal_weights.shape[0]
+            != self.acq_func.model.train_inputs[0][-1].shape[-1]
+        ):
+            raise ValueError(
+                "`proximal_weights` must be a one dimensional tensor with "
+                "same feature dimension as model."
+            )
+
+    @t_batch_mode_transform(expected_q=1, assert_output_shape=False)
+    def forward(self, X: Tensor) -> Tensor:
+        r"""Evaluate base acquisition function with proximal weighting.
+
+        Args:
+            X: Input tensor of feature dimension `d` .
+
+        Returns:
+            Base acquisition function evaluated on tensor `X` multiplied by proximal
+            weighting.
+        """
+        d = self.proximal_weights.shape[0]
+        last_X = self.acq_func.model.train_inputs[0][-1].reshape(1, 1, -1)
+        diff = X - last_X
+        L = torch.diag(torch.sqrt(self.proximal_weights)).reshape(1, d, d)
+
+        M = _batch_mahalanobis(L, diff)
```

Comment:
I still don't think using `_batch_mahalanobis` is necessary here. Since the weight matrix is diagonal by construction, there is no need to do fancy triangular solves or anything since this simplifies to the following. 
```suggestion
        M = (diff ** 2 / self.proximal_weights).sum()
```